model/DataObject.php

```diff
@@ -1876,168 +1931,241 @@ public function has_many($component = null, $classOnly = true) {
+		$hasMany = (array)Config::inst()->get($this->class, 'has_many', Config::INHERITED);
+
+		if($hasMany && array_key_exists($component, $hasMany)) {
+			$hasMany = $hasMany[$component];
+		} else {
+			return false;
+		}
+
+		return ($classOnly) ? preg_replace('/(.+)?\..+/', '$1', $hasMany) : $hasMany;
+	}
+
+	/**
+	 * @deprecated 4.0 Method has been renamed to manyManyExtraFields()
+	 * @param string $component
+	 * @return array
+	 */
+	public function many_many_extraFields($component = null) {
+		Deprecation::notice('4.0', 'Please use manyManyExtraFields() instead');
+		return $this->manyManyExtraFields($component);
+	}
+
+	/**
 	 * Return the many-to-many extra fields specification.
 	 *
 	 * If you don't specify a component name, it returns all
 	 * extra fields for all components available.
 	 *
 	 * @param string $component Name of component
-	 * @return array
+	 * @return array|null
 	 */
-	public function many_many_extraFields($component = null) {
-		$classes = ClassInfo::ancestry($this);
-
-		foreach($classes as $class) {
-			if(in_array($class, array('ViewableData', 'Object', 'DataObject'))) continue;
-			$relationName = null;
-
-			// Find extra fields for one component
-			if($component) {
-				$SNG_class = singleton($class);
-				$extraFields = $SNG_class->stat('many_many_extraFields');
-
-				// Extra fields are immediately available on this class
-				if(isset($extraFields[$component])) {
-					return $extraFields[$component];
-				}
+	public function manyManyExtraFields($component = null) {
+		if($component) {
+			return $this->manyManyExtraFieldsForComponent($component);
+		}
 
-				$manyMany = $SNG_class->stat('many_many');
-				$candidate = (isset($manyMany[$component])) ? $manyMany[$component] : null;
-				if($candidate) {
-					$SNG_candidate = singleton($candidate);
-					$candidateManyMany = $SNG_candidate->stat('belongs_many_many');
+		$items = (array)Config::inst()->get($this->class, 'many_many_extraFields', Config::INHERITED);
+		foreach($items as $k => $v) {
+			if(!is_array($v)) {
+				throw new LogicException("$this->class::\$many_many_extraFields has a bad entry: "
+					. var_export($k, true) . " => " . var_export($v, true)
+					. ". Each many_many_extraFields entry should map to a field specification array.");
+			}
+		}
 
-					// Find the relation given the class
-					if($candidateManyMany) foreach($candidateManyMany as $relation => $relatedClass) {
-						if($relatedClass == $class) {
-							$relationName = $relation;
-							break;
-						}
-					}
+		return count($items) ? $items : null;
+	}
 
-					if($relationName) {
-						$extraFields = $SNG_candidate->stat('many_many_extraFields');
-						if(isset($extraFields[$relationName])) {
-							return $extraFields[$relationName];
-						}
-					}
+	/**
+	 * Return the many-to-many extra fields specification for a specific component.
+	 * @param string $component
+	 * @return array|null
+	 */
+	public function manyManyExtraFieldsForComponent($component) {
```

Comment:
I think this method is taking on too much responsibility. Each bit where you're looping can probably be it's own function.

Also, isn't the logic for extracting the reverse relation via dot notation a share-able set of logic that could be a function too?

The fact you're returning in many different places under different conditions shows that it's a good candidate for being split into component parts